src/Microsoft.ML.Transforms/PermutationFeatureImportance.cs

```diff
@@ -0,0 +1,212 @@
+                        nextFeatureIndex = workingFeatureIndices[j + 1];
+
+                    int nextValuesIndex = 0;
+
+                    Utils.Shuffle(RandomUtils.Create(shuffleSeed), featureValuesBuffer);
+
+                    Action<FeaturesBuffer, FeaturesBuffer, PermuterState> permuter =
+                        (src, dst, state) =>
+                        {
+                            src.Features.CopyTo(ref dst.Features);
+                            VBufferUtils.ApplyAt(ref dst.Features, workingIndx,
+                                (int ii, ref float d) =>
+                                    d = featureValuesBuffer[state.SampleIndex++]);
+
+                            if (j < workingFeatureIndices.Count - 1)
+                            {
+                                // This is the reason I need PermuterState in LambdaTransform.CreateMap.
+                                nextValues[nextValuesIndex] = src.Features.GetItemOrDefault(nextFeatureIndex);
+                                if (nextValuesIndex < valuesRowCount - 1)
+                                    nextValuesIndex++;
+                            }
+                        };
+
+                    SchemaDefinition input = SchemaDefinition.Create(typeof(FeaturesBuffer));
+                    Contracts.Assert(input.Count == 1);
+                    input[0].ColumnName = features;
+
+                    SchemaDefinition output = SchemaDefinition.Create(typeof(FeaturesBuffer));
+                    Contracts.Assert(output.Count == 1);
+                    output[0].ColumnName = features;
+                    output[0].ColumnType = featuresColumn.Type;
+
+                    IDataView viewPermuted = LambdaTransform.CreateMap(
+                        host, data, permuter, null, input, output);
+                    if (topExamples > 0 && valuesRowCount == topExamples)
+                        viewPermuted = SkipTakeFilter.Create(host, new SkipTakeFilter.TakeArguments() { Count = valuesRowCount }, viewPermuted);
+
+                    var metrics = evaluationFunc(model.Transform(viewPermuted));
+
+                    var delta = deltaFunc(metrics, baselineMetrics);
+                    metricsDelta.Add(delta);
+
+                    // Swap values for next iteration of permutation.
+                    Array.Clear(featureValuesBuffer, 0, featureValuesBuffer.Length);
+                    nextValues.CopyTo(featureValuesBuffer, 0);
+                    Array.Clear(nextValues, 0, nextValues.Length);
+
+                    // Print out timings.
+                    if (processedCnt > 0 && (processedCnt % progressIterations == 0))
+                    {
+                        stopwatch.Stop();
+                        ch.Info(string.Format("Processed slots {0} - {1} in {2}", processedCnt - progressIterations,
+                            processedCnt, stopwatch.Elapsed));
+                        stopwatch.Restart();
+                    }
```

Comment:
We have progress channel for this.
Please don't use stopwatch.

[https://github.com/dotnet/machinelearning/blob/0fe8bafa11e0b759425100db12327569a8c444bb/src/Microsoft.ML.StandardLearners/Standard/SdcaBinary.cs#L344](https://github.com/dotnet/machinelearning/blob/0fe8bafa11e0b759425100db12327569a8c444bb/src/Microsoft.ML.StandardLearners/Standard/SdcaBinary.cs#L344)
here is example of how it should be done.

just set header with one metric: processed slots, use SetProgress without limit, and set checkpoint after your loop.